File: assets/js/dispatcher.js
export class Dispatcher {
    constructor() {
        this.listeners = new Map();
    }

    addListener(eventName, listener) {
        if (!this.listeners.has(eventName)) {
            this.listeners.set(eventName, []);
        }

        let eventListeners = this.listeners.get(eventName);
        eventListeners.push(listener);

        this.listeners.set(eventName, eventListeners);
    }

    dispatch(eventName, parameters) {
        if (!this.listeners.has(eventName)) {
            return null;
        }

        let result = null;
        for (let listener of this.listeners.get(eventName)) {
            result = listener(parameters);
        }

        return result;
    }
}

// export const mainDispatcher = new Dispatcher()

File: pkg/util/dispatcher.go
package util

import (
	"slices"
	"sync"
)

type Event interface {
	GetName() string
}

type EventHandler func(Event)

type Dispatcher struct {
	mu        sync.RWMutex
	listeners map[string][]EventHandler
}

func NewDispatcher() *Dispatcher {
	return &Dispatcher{
		listeners: make(map[string][]EventHandler),
	}
}

func (d *Dispatcher) AddListener(eventName string, handler EventHandler) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.listeners[eventName] = append(d.listeners[eventName], handler)
}

func (d *Dispatcher) Dispatch(event Event) {
	d.mu.RLock()
	handlers := slices.Clone(d.listeners[event.GetName()])
	d.mu.RUnlock()
	for _, handler := range handlers {
		handler(event)
	}
}
